pkg/sql/plan2/function/function.go

```diff
@@ -0,0 +1,277 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package function
+
+import (
+	"fmt"
+	"github.com/matrixorigin/matrixone/pkg/container/nulls"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/encoding"
+	"github.com/matrixorigin/matrixone/pkg/errno"
+	"github.com/matrixorigin/matrixone/pkg/pb/plan"
+	"github.com/matrixorigin/matrixone/pkg/sql/errors"
+	"github.com/matrixorigin/matrixone/pkg/vectorize/add"
+	"github.com/matrixorigin/matrixone/pkg/vm/process"
+	"reflect"
+	"sync"
+)
+
+const (
+	// NoLimit is a flag that means no number limit for argument number
+	NoLimit = -1
+
+	// NullValueType is the type of const value `NULL`
+	// If input NULL as an argument, its type should be NullValue which can match each type.
+	// e.g.
+	// built_in_function(100, NULL)
+	// its argument type is [types.T_int64, NullValueType]
+	NullValueType = types.T_any
+)
+
+var emptyFunction = Function{Name: "empty function structure"}
+
+// Function is an overload of
+// a built-in function or an aggregate function or an operator
+type Function struct {
+	// Name records info of the function overload used to print
+	Name string
+	Flag plan.Function_FuncFlag
+
+	Args      ArgList
+	ReturnTyp types.T
+
+	// ID is a unique flag for the overload. It is helpful to find the overload quickly.
+	ID int64
+
+	// Fn is executed method of built-in function and operator
+	Fn func(vs []*vector.Vector, proc *process.Process) (*vector.Vector, error)
+
+	// AggregateInfo is related information if it's an aggregate function.
+	// TODO(cms): maybe it will record the Ring id
+	AggregateInfo interface{}
+
+	// SQLFn returns the sql string of the function. Maybe useful.
+	// TODO(cms): if useless, just remove it.
+	SQLFn func(vs []*vector.Vector) (string, error)
+}
+
+// ArgList is a structure to record argument info of a function overload.
+type ArgList struct {
+	Limit bool // if true, argument number of function is constant
+
+	// if Limit is true, ArgTypes1 records all argument types
+	ArgTypes1 []Arg
+
+	// if Limit is false, ArgTypes2 records all argument types
+	ArgTypes2 struct {
+		// TypeCheckFn is function's own argument type check function.
+		TypeCheckFn func(inputTypes []types.T) bool
+	}
+}
+
+type Arg struct {
+	Name string // for function help, or for developers
+	Typ  types.T
+}
+
+var registerMutex = sync.RWMutex{} // read/write lock for functionRegister
+var functionRegister = map[string][]Function{
+	// Operators
+	"=": []Function{
+		{
+			Name: "=(uint8, uint8)",
+			Flag: plan.Function_STRICT,
+			Args: MakeLimitArgList([]Arg{
+				{Name: "left", Typ: types.T_uint8},
```

Comment:
SQL arg is positional, so the Name here is useless.

Reply:
you're right. 
I just records it when I was writing the code and I didn't find the use of it.

My initial thought was maybe we can print some error message using that.